[View in Colaboratory](https://colab.research.google.com/github/boluodehaiwangzi/voice-enabled-interactive-learning/blob/gh-pages/deep_understand_mnist.ipynb)

# 整个MNITST 项目是分三个阶段进行的：
## 1、简单的全连接神经网络模型，准确率可以达到92%
## 2、全连接模型使用：relu，dropout，adam函数，准确率98%
## 3、卷积神经网络 99.2%

In [4]:
# 第一个项目，先实现传统的全连接神经网络

import tensorflow as tf
mnist = tf.keras.datasets.mnist

# 导入数据，并将数据转变到0,1之间
(x_train,y_train),(x_test,y_test) = mnist.load_data()
x_train,x_test = x_train/255.0, x_test/255.0

# 定义模型的各个层

model = tf.keras.models.Sequential([
    tf.keras.layers.Flatten(),
    tf.keras.layers.Dense(512,activation=tf.nn.sigmoid),
    tf.keras.layers.Dense(10,activation=tf.nn.softmax)
]
)

# 定义的优化器，损失函数
model.compile(optimizer = 'adam',loss='sparse_categorical_crossentropy',metrics=['accuracy'])

# 训练
model.fit(x_train, y_train,epochs=30)
model.evaluate(x_test,y_test)


# 结论：
# 1、使用以上方式代码可以实现92.2%的准确率，对于SGD优化器来说，学习速度偏慢，所以需要迭代的次数也比较多，10次迭代还没收敛，需要25次才开始收敛
# 2、在没有改变其他参数的情况下，仅仅改变优化器成adam，训练准确率可达99.99%，测试准确率98.2%

Epoch 1/30
60000/60000 [==============================] - 9s 147us/step - loss: 0.3398 - acc: 0.9033
Epoch 2/30
60000/60000 [==============================] - 9s 143us/step - loss: 0.1698 - acc: 0.9494
Epoch 3/30
60000/60000 [==============================] - 9s 144us/step - loss: 0.1124 - acc: 0.9665
Epoch 4/30
60000/60000 [==============================] - 9s 143us/step - loss: 0.0783 - acc: 0.9765
Epoch 5/30
60000/60000 [==============================] - 9s 143us/step - loss: 0.0568 - acc: 0.9835
Epoch 6/30
60000/60000 [==============================] - 9s 143us/step - loss: 0.0423 - acc: 0.9872
Epoch 7/30
60000/60000 [==============================] - 9s 145us/step - loss: 0.0316 - acc: 0.9909
Epoch 8/30
60000/60000 [==============================] - 9s 144us/step - loss: 0.0236 - acc: 0.9933
Epoch 9/30
60000/60000 [==============================] - 9s 144us/step - loss: 0.0174 - acc: 0.9954
Epoch 10/30
60000/60000 [==============================] - 9s 145us/step - loss: 0.0131 - a

[0.07959315801936609, 0.9828]

In [8]:
# 使用relu激活函数代替sigmoid函数，dropout，adam优化器

import tensorflow as tf
mnist = tf.keras.datasets.mnist

# 导入数据
(x_train,y_train),(x_test,y_test) = mnist.load_data()
x_train, x_test = x_train/255.0, x_test/255.0

# 定义模型
model = tf.keras.models.Sequential([
    tf.keras.layers.Flatten(),
    tf.keras.layers.Dense(512,activation='relu'),
    tf.keras.layers.Dropout(0.2),
    tf.keras.layers.Dense(10,activation='softmax')
])

# 设定优化器,损失函数，输出结果方式
model.compile(optimizer='adam',loss='sparse_categorical_crossentropy',metrics=['accuracy'])

# 训练
model.fit(x_train,y_train,epochs=10)
model.evaluate(x_test,y_test)

# 总结：
# 1、学习效果很好，五次训练就可以达到98.6%的效果，并在测试集上达到98%的效果
# 2、adam优化器在训练集上，随着训练次数增加，训练集上的准确率会提升，但是测试集上的准确率不会
# 3、adam优化器可能比较容易过拟合

Epoch 1/10
60000/60000 [==============================] - 9s 158us/step - loss: 0.2198 - acc: 0.9348
Epoch 2/10
60000/60000 [==============================] - 9s 156us/step - loss: 0.0972 - acc: 0.9700
Epoch 3/10
60000/60000 [==============================] - 9s 155us/step - loss: 0.0691 - acc: 0.9781
Epoch 4/10
60000/60000 [==============================] - 9s 155us/step - loss: 0.0524 - acc: 0.9842
Epoch 5/10
60000/60000 [==============================] - 9s 154us/step - loss: 0.0447 - acc: 0.9852
Epoch 6/10
60000/60000 [==============================] - 9s 155us/step - loss: 0.0369 - acc: 0.9880
Epoch 7/10
60000/60000 [==============================] - 9s 157us/step - loss: 0.0321 - acc: 0.9892
Epoch 8/10
60000/60000 [==============================] - 9s 155us/step - loss: 0.0256 - acc: 0.9915
Epoch 9/10
60000/60000 [==============================] - 9s 156us/step - loss: 0.0254 - acc: 0.9914
Epoch 10/10
10000/10000 [==============================] - 1s 57us/step


[0.07637497371167191, 0.9817]

In [1]:
# 例程：类似VGG的卷积神经网络实现

import numpy as np
import keras
from keras.models import Sequential
from keras.layers import Dense, Dropout, Flatten
from keras.layers import Conv2D, MaxPooling2D
from keras.optimizers import SGD

# Generate dummy data
x_train = np.random.random((100, 100, 100, 3))
y_train = keras.utils.to_categorical(np.random.randint(10, size=(100, 1)), num_classes=10)
x_test = np.random.random((20, 100, 100, 3))
y_test = keras.utils.to_categorical(np.random.randint(10, size=(20, 1)), num_classes=10)

model = Sequential()
# input: 100x100 images with 3 channels -> (100, 100, 3) tensors.
# this applies 32 convolution filters of size 3x3 each.
model.add(Conv2D(32, (3, 3), activation='relu', input_shape=(100, 100, 3)))
model.add(Conv2D(32, (3, 3), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))

model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))

model.add(Flatten())
model.add(Dense(256, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(10, activation='softmax'))

sgd = SGD(lr=0.01, decay=1e-6, momentum=0.9, nesterov=True)
model.compile(loss='categorical_crossentropy', optimizer=sgd)

model.fit(x_train, y_train, batch_size=32, epochs=10)
score = model.evaluate(x_test, y_test, batch_size=32)


Using TensorFlow backend.


Epoch 1/10
100/100 [==============================] - 6s 65ms/step - loss: 2.3396
Epoch 2/10
100/100 [==============================] - 5s 55ms/step - loss: 2.2968
Epoch 3/10
100/100 [==============================] - 5s 54ms/step - loss: 2.2581
Epoch 4/10
100/100 [==============================] - 5s 54ms/step - loss: 2.2869
Epoch 5/10
100/100 [==============================] - 5s 54ms/step - loss: 2.2851
Epoch 6/10
100/100 [==============================] - 5s 54ms/step - loss: 2.2520
Epoch 7/10
100/100 [==============================] - 5s 54ms/step - loss: 2.2467
Epoch 8/10
100/100 [==============================] - 5s 54ms/step - loss: 2.2640
Epoch 9/10
100/100 [==============================] - 5s 54ms/step - loss: 2.2236
Epoch 10/10
20/20 [==============================] - 0s 17ms/step


# CNN MNIST实现

In [9]:
import tensorflow as tf
import keras
from keras.models import Sequential
from keras.layers import Dense, Dropout, Flatten
from keras.layers import Conv2D, MaxPooling2D

(x_train,y_train),(x_test,y_test) = keras.datasets.mnist.load_data()
x_train_tem, x_test_tem = x_train/255.0, x_test/255.0
x_train = x_train_tem.reshape(60000,28,28,1)
x_test = x_test_tem.reshape(10000,28,28,1)

model = Sequential()
model.add(Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)))
#model.add(Conv2D(32, (3, 3), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.2))

#model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.2))

model.add(Flatten())
model.add(Dense(512, activation='relu'))
model.add(Dropout(0.2))
model.add(Dense(10, activation='softmax'))

# 优化器，目标函数，输出形式
model.compile(optimizer='adam',loss='sparse_categorical_crossentropy',metrics=['accuracy'])

# 训练
model.fit(x_train,y_train,epochs=5)
model.evaluate(x_test,y_test)

Epoch 1/5
60000/60000 [==============================] - 97s 2ms/step - loss: 0.1468 - acc: 0.9533
Epoch 2/5
60000/60000 [==============================] - 98s 2ms/step - loss: 0.0551 - acc: 0.9827
Epoch 3/5
60000/60000 [==============================] - 96s 2ms/step - loss: 0.0400 - acc: 0.9873
Epoch 4/5
60000/60000 [==============================] - 97s 2ms/step - loss: 0.0341 - acc: 0.9894
Epoch 5/5
10000/10000 [==============================] - 4s 434us/step


[0.02310111620900716, 0.9925]

Epoch 1/5
60000/60000 [==============================] - 97s 2ms/step - loss: 0.1468 - acc: 0.9533
Epoch 2/5
60000/60000 [==============================] - 98s 2ms/step - loss: 0.0551 - acc: 0.9827
Epoch 3/5
60000/60000 [==============================] - 96s 2ms/step - loss: 0.0400 - acc: 0.9873
Epoch 4/5
60000/60000 [==============================] - 97s 2ms/step - loss: 0.0341 - acc: 0.9894
Epoch 5/5
60000/60000 [==============================] - 98s 2ms/step - loss: 0.0273 - acc: 0.9913
10000/10000 [==============================] - 4s 434us/step
[0.02310111620900716, 0.9925]